Dockerfile-base-cpu

```diff
@@ -41,6 +44,18 @@ ENV PYTHONUNBUFFERED=1 PYTHONFAULTHANDLER=1 PYTHONHASHSEED=0
 ARG PYTHON_VERSION
 RUN /tmp/det_dockerfile_scripts/install_python.sh ${PYTHON_VERSION}
 
+ARG USE_MPI
+ARG UCX_INSTALL_DIR=/container/ucx
+ARG OMPI_INSTALL_DIR=/container/ompi
+RUN if [ "$USE_MPI" = "1" ]; then /tmp/det_dockerfile_scripts/ompi.sh; fi
+# Make sure OMPI/UCX show up in the right paths
+ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:/usr/lib/libibverbs:${UCX_INSTALL_DIR}/lib:${UCX_INSTALL_DIR}/lib64:${OMPI_INSTALL_DIR}/lib
+ENV PATH ${UCX_INSTALL_DIR}/bin:${OMPI_INSTALL_DIR}/bin:$PATH
+# Enable running OMPI as root
+ENV OMPI_ALLOW_RUN_AS_ROOT=1 \
+    OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
```

Comment:
Is there a good way to do that?  I didn't know a good way to do a conditional check within the Dockerfile that would also correctly export the ENV variable in the final image.  Happy to change this if you have an example.

Reply:
I have a similar code here, which only sets `HOROVOD_NCCL_HOME` to `/tmp/det_nccl....`  when `DET_BUILD_NCCL` is enabled: https://github.com/determined-ai/environments/blob/master/Dockerfile-default-gpu#L33

Reply:
Doesn't that just set the variable for the docker build?  The ENV is to set the variable in the environment of the running container.  I didn't know that ARG settings persisted into the environment of the final container.  Is that true?

Reply:
I looked more into this and my understanding seems right that ARG values are for the build step and do not persist in the final image.  I also found that you cannot export the value within a script and have that persist to subsequent steps or the final image.  It seems to me that the only option is to always set these using ENV.  I could either set them to 1 or 0, depending on if USE_MPI is set or not, but I think they must be set to something regardless.

Do you have an opinion or some solution I'm missing?  Would you prefer the values be set based on USE_MPI so they are either 1 or 0 but always there?

Reply:
My point was about ${VAR:+value} substitution, not so much ENV vs ARG.
Overall I think it'd be great to avoid having PATH as `${UCX_INSTALL_DIR}/bin:${OMPI_INSTALL_DIR}/bin:$PATH` where first two directories are missing in non-mpi images.

Here's an example:
```
FROM ubuntu:20.04

ARG DET_USE_MPI
ARG UCX_INSTALL_DIR=/container/ucx
ENV OMPI_ALLOW_RUN_AS_ROOT ${DET_USE_MPI:+1}
ENV DET_PATH_PREFIX ${DET_USE_MPI:+$UCX_INSTALL_DIR:}
ENV PATH $DET_PATH_PREFIX$PATH
```

With this, `docker build -f Dockerfile -t test-condos . ; docker run -it test-condos /bin/bash -c env` will have the original PATH. 
`docker build -f Dockerfile -t test-condos --build-arg DET_USE_MPI=1 . ; docker run -it test-condos /bin/bash -c env` will have an updated PATH and `OMPI_ALLOW_RUN_AS_ROOT=1`.

Caveat: in this example, `DET_PATH_PREFIX` has to be `ENV` for some reason and it ends up leaking into the container as an variable. ideally it'd be `ARG`, but that doesn't work for me for some reason.

Reply:
I think I got this working as close as possible to what you're suggesting.  Here is what is in the Dockerfile now:
```
ARG USE_MPI
ARG UCX_INSTALL_DIR=/container/ucx
ARG OMPI_INSTALL_DIR=/container/ompi
RUN if [ "$USE_MPI" = "1" ]; then /tmp/det_dockerfile_scripts/ompi.sh; fi
# Make sure OMPI/UCX show up in the right paths                                 
ARG VERBS_LIB_DIR=/usr/lib/libibverbs
ARG UCX_LIB_DIR=${UCX_INSTALL_DIR}/lib:${UCX_INSTALL_DIR}/lib64
ARG UCX_PATH_DIR=${UCX_INSTALL_DIR}/bin
ARG OMPI_LIB_DIR=${OMPI_INSTALL_DIR}/lib
ARG OMPI_PATH_DIR=${OMPI_INSTALL_DIR}/bin
ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:${USE_MPI:+$VERBS_LIB_DIR:$UCX_LIB_DIR:$OM\
PI_LIB_DIR}
ENV PATH $PATH:${USE_MPI:+$UCX_PATH_DIR:$OMPI_PATH_DIR}
# Enable running OMPI as root                                                   
ENV OMPI_ALLOW_RUN_AS_ROOT ${USE_MPI:+1}
ENV OMPI_ALLOW_RUN_AS_ROOT_CONFIRM ${USE_MPI:+1}
```
Now if I do not specify USE_MPI=1 on the make line I see the following in the container:
```
root@0f5057397c59:/# env | grep OMPI
OMPI_ALLOW_RUN_AS_ROOT=
OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=
root@0f5057397c59:/# echo $PATH
/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:
root@0f5057397c59:/# echo $LD_LIBRARY_PATH
:
```
Note that the OMPI* variables still show up but they are empty.  I couldn't find a way to have them not exist at all if USE_MPI was not specified.

With USE_MPI=1 passed to the make line I get:
```
root@40bc3fec58c8:/# env | grep OMPI
OMPI_ALLOW_RUN_AS_ROOT=1
OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
root@40bc3fec58c8:/# echo $PATH
/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/container/ucx/bin:/container/ompi/bin
root@40bc3fec58c8:/# echo $LD_LIBRARY_PATH
:/usr/lib/libibverbs:/container/ucx/lib:/container/ucx/lib64:/container/ompi/lib
```
To make this work I modified the Makefile to define a USE_MPI=1 or USE_GLOO=1 and this is given as a build arg so the Dockerfile knows whether to set the env or not.